[Automation] java run fails for org.eclipse.jetty:jetty-alpn-client:12.1.0

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25541257498/job/74967785015
Last 300 lines of the log:
```console
Testing library: org.eclipse.jetty:jetty-alpn-client:12.0.31
Command: `/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/gradlew nativeTest`
Executing test...
-------
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :jar
> Task :compileTestJava
> Task :checkTestTimeoutAnnotations
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@28f3e4f3::oeji.ByteArrayEndPoint@2450de9[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=9/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@28f3e4f3][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@3637e0c[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@4f6202fd,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)

10 tests completed, 2 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

BUILD FAILED in 3s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.eclipse.jetty:jetty-alpn-client:12.0.31 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 12.1.0 at 2026-05-08T06:53:13Z with exit code 1 after 5s
❌ ERROR: 'native-image run' for 12.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean test -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 12.1.0 at 2026-05-08T06:53:13Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 797ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 984ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 12.1.0 at 2026-05-08T06:53:16Z with exit code 0 after 3s
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage java run
========================================================================================
Starting stage 'java run' for 12.1.0 at 2026-05-08T06:53:16Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 815ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processTestResources NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :testClasses

> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@4a22cf26::oeji.ByteArrayEndPoint@515fcdaf[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=7/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@4a22cf26][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@6643f600[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@25cd000b,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)

10 tests completed, 2 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.


FAILURE: Build failed with an exception.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date
* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 12.1.0 at 2026-05-08T06:53:21Z with exit code 1 after 5s
❌ ERROR: 'java run' for 12.1.0 failed with exit code 1.
FAILED[java run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25541257498/job/74967785232
Last 300 lines of the log:
```console
Testing library: org.eclipse.jetty:jetty-alpn-client:12.0.31
Command: `/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/gradlew nativeTest`
Executing test...
-------
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :jar
> Task :checkTestTimeoutAnnotations
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@72f68732::oeji.ByteArrayEndPoint@7d797ad9[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=6/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@72f68732][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@53dab18[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@419569d4,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)

10 tests completed, 2 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).


BUILD FAILED in 3s
14 actionable tasks: 4 executed, 10 up-to-date

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.eclipse.jetty:jetty-alpn-client:12.0.31 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 12.1.0 at 2026-05-08T06:52:50Z with exit code 1 after 5s
❌ ERROR: 'native-image run' for 12.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean test -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 12.1.0 at 2026-05-08T06:52:50Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 773ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 12.1.0 at 2026-05-08T06:52:53Z with exit code 0 after 3s
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage java run
========================================================================================
Starting stage 'java run' for 12.1.0 at 2026-05-08T06:52:53Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 744ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :processTestResources NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :testClasses

> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null

10 tests completed, 2 failed
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED

FAILURE: Build failed with an exception.

* What went wrong:    java.lang.AssertionError: 

Execution failed for task ':test'.
>     Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@194ebefa::oeji.ByteArrayEndPoint@31d9e284[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=6/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@194ebefa][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@67fb83bb[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@46822375,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.htmlThere were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).


Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.


BUILD FAILED in 2s
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 12.1.0 at 2026-05-08T06:52:57Z with exit code 1 after 4s
❌ ERROR: 'java run' for 12.1.0 failed with exit code 1.
FAILED[java run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25541257498/job/74967785222
Last 300 lines of the log:
```console
Testing library: org.eclipse.jetty:jetty-alpn-client:12.0.31
Command: `/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/gradlew nativeTest`
Executing test...
-------
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :jar
> Task :checkTestTimeoutAnnotations
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@354868fd::oeji.ByteArrayEndPoint@21115b80[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=7/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@354868fd][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@595bea64[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@189f4202,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

10 tests completed, 2 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

BUILD FAILED in 3s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.eclipse.jetty:jetty-alpn-client:12.0.31 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 5s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 12.1.0 at 2026-05-08T06:53:36Z with exit code 1 after 6s
❌ ERROR: 'native-image run' for 12.1.0 failed with exit code 1.
SOFT-FAIL[native-image run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="12.1.0" ./gradlew clean test -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 12.1.0 at 2026-05-08T06:53:36Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="12.1.0" ./gradlew clean compileTestJava -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 817ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1s
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 12.1.0 at 2026-05-08T06:53:39Z with exit code 0 after 3s
========================================================================================
 org.eclipse.jetty:jetty-alpn-client:12.0.31:12.1.0 stage java run
========================================================================================
Starting stage 'java run' for 12.1.0 at 2026-05-08T06:53:39Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 855ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :processTestResources NO-SOURCE
> Task :testClasses


> Task :test FAILED

Jetty_alpn_clientTest > failedProtocolConnectionCreationClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting actual not to be null
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.failedProtocolConnectionCreationClosesEndPoint(Jetty_alpn_clientTest.java:176)

Jetty_alpn_clientTest > endOfInputFailsConnectionPromiseAndClosesEndPoint() FAILED
    java.lang.AssertionError: 
    Expecting empty but was: [oj.Jetty_alpn_clientTest$RecordingConnection@6f12596e::oeji.ByteArrayEndPoint@6fadb38d[{l=/0.0.0.0:0,r=/0.0.0.0:0,ISHUT,fill=-,flush=-,to=7/0}]->[oj.Jetty_alpn_clientTest$RecordingConnection@6f12596e][q=1,q[0]=java.nio.HeapByteBuffer[pos=0 lim=0 cap=0],o=oeji.RetainableByteBuffer$DynamicCapacity@101dfd2c[0/9223372036854775807,d=false,pool=org.eclipse.jetty.io.ByteBufferPool$Sized@2c119c6,minRetain=-1,buffers=0,rc=1]]]
        at org_eclipse_jetty.jetty_alpn_client.Jetty_alpn_clientTest.endOfInputFailsConnectionPromiseAndClosesEndPoint(Jetty_alpn_clientTest.java:138)
10 tests completed, 2 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 2s

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.eclipse.jetty/jetty-alpn-client/12.0.31/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 4s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 12.1.0 at 2026-05-08T06:53:43Z with exit code 1 after 4s
❌ ERROR: 'java run' for 12.1.0 failed with exit code 1.
FAILED[java run][12.1.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="12.1.0" ./gradlew clean javaTest -Pcoordinates="org.eclipse.jetty:jetty-alpn-client:12.0.31"]
```

--- a/metadata/org.eclipse.jetty/jetty-alpn-client/12.1.0/reachability-metadata.json
+++ b/metadata/org.eclipse.jetty/jetty-alpn-client/12.1.0/reachability-metadata.json
@@ -1,0 +1,30 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "org.eclipse.jetty.io.IdleTimeout"
+      },
+      "type": "java.lang.StackTraceElement[]"
+    },
+    {
+      "condition": {
+        "typeReached": "org.eclipse.jetty.util.TypeUtil"
+      },
+      "type": "java.lang.invoke.MethodHandle[]"
+    }
+  ],
+  "resources": [
+    {
+      "condition": {
+        "typeReached": "org.eclipse.jetty.util.ServiceLoaderSpliterator"
+      },
+      "glob": "META-INF/services/org.eclipse.jetty.io.ssl.ALPNProcessor$Client"
+    },
+    {
+      "condition": {
+        "typeReached": "org.eclipse.jetty.io.IdleTimeout"
+      },
+      "glob": "META-INF/services/org.slf4j.spi.SLF4JServiceProvider"
+    }
+  ]
+}

--- a/metadata/org.eclipse.jetty/jetty-alpn-client/index.json
+++ b/metadata/org.eclipse.jetty/jetty-alpn-client/index.json
@@ -2,6 +2,11 @@
   {
     "latest" : true,
     "metadata-version" : "12.1.0",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-alpn-client/$version$/jetty-alpn-client-$version$-sources.jar",
+    "repository-url" : "https://github.com/jetty/jetty.project",
+    "test-code-url" : "N/A",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-alpn-client/$version$/jetty-alpn-client-$version$-javadoc.jar",
+    "description" : "Jetty ALPN Client provides the client-side connection factory used by Eclipse Jetty to negotiate application protocols such as HTTP/1.1 and HTTP/2 during TLS handshakes. It discovers available Jetty ALPN processors and configures SSL client connections so the selected protocol can hand off to the appropriate client connection implementation.",
     "tested-versions" : [
       "12.1.0"
     ],

--- a/metadata/org.eclipse.jetty/jetty-alpn-client/index.json
+++ b/metadata/org.eclipse.jetty/jetty-alpn-client/index.json
@@ -1,6 +1,17 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "12.1.0",
+    "tested-versions" : [
+      "12.1.0"
+    ],
+    "allowed-packages" : [
+      "org.eclipse.jetty.alpn.client",
+      "org.eclipse.jetty.util",
+      "org.eclipse.jetty.io"
+    ]
+  },
+  {
     "metadata-version" : "12.0.31",
     "source-code-url" : "https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-alpn-client/$version$/jetty-alpn-client-$version$-sources.jar",
     "repository-url" : "https://github.com/jetty/jetty.project",

--- a/stats/org.eclipse.jetty/jetty-alpn-client/12.1.0/execution-metrics.json
+++ b/stats/org.eclipse.jetty/jetty-alpn-client/12.1.0/execution-metrics.json
@@ -1,0 +1,87 @@
+{
+  "fix_java_run_fail:2026-06-06": {
+    "timestamp": "2026-06-06T06:13:12.128480Z",
+    "library": "org.eclipse.jetty:jetty-alpn-client:12.1.0",
+    "starting_commit": "dfb43e8f5b53aff88a17b90cb0b0dc9904d9d494",
+    "ending_commit": "42c2ca80a3e0ee525adb615020a5df48216e633c",
+    "previous_library": "org.eclipse.jetty:jetty-alpn-client:12.0.31",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "12.1.0",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 78,
+          "missed": 119,
+          "ratio": 0.395939,
+          "total": 197
+        },
+        "line": {
+          "covered": 21,
+          "missed": 28,
+          "ratio": 0.428571,
+          "total": 49
+        },
+        "method": {
+          "covered": 7,
+          "missed": 2,
+          "ratio": 0.777778,
+          "total": 9
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "12.0.31",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 65,
+          "missed": 118,
+          "ratio": 0.355191,
+          "total": 183
+        },
+        "line": {
+          "covered": 18,
+          "missed": 28,
+          "ratio": 0.391304,
+          "total": 46
+        },
+        "method": {
+          "covered": 5,
+          "missed": 2,
+          "ratio": 0.714286,
+          "total": 7
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 47981,
+      "cached_input_tokens_used": 392192,
+      "output_tokens_used": 4280,
+      "iterations": 1,
+      "cost_usd": 0.3245,
+      "tested_library_loc": 21,
+      "metadata_entries": 4,
+      "previous_library_metadata_entries": 4,
+      "code_coverage_percent": 42.86,
+      "previous_library_coverage_percent": 39.13
+    },
+    "artifacts": {
+      "test_file": "tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/src/test/java/org_eclipse_jetty/jetty_alpn_client/Jetty_alpn_clientTest.java",
+      "metadata_file": "metadata/org.eclipse.jetty/jetty-alpn-client/12.1.0/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/org.eclipse.jetty/jetty-alpn-client/12.1.0/stats.json
+++ b/stats/org.eclipse.jetty/jetty-alpn-client/12.1.0/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "12.1.0",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 78,
+        "missed" : 119,
+        "ratio" : 0.395939,
+        "total" : 197
+      },
+      "line" : {
+        "covered" : 21,
+        "missed" : 28,
+        "ratio" : 0.428571,
+        "total" : 49
+      },
+      "method" : {
+        "covered" : 7,
+        "missed" : 2,
+        "ratio" : 0.777778,
+        "total" : 9
+      }
+    }
+  } ]
+}

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/.gitignore
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/build.gradle
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/build.gradle
@@ -1,0 +1,27 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.eclipse.jetty:jetty-alpn-client:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/gradle.properties
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = org.eclipse.jetty:jetty-alpn-client:12.1.0
+library.version = 12.1.0
+metadata.dir = org.eclipse.jetty/jetty-alpn-client/12.1.0/

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/settings.gradle
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'org.eclipse.jetty.jetty-alpn-client_tests'

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/src/test/java/org_eclipse_jetty/jetty_alpn_client/Jetty_alpn_clientTest.java
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/src/test/java/org_eclipse_jetty/jetty_alpn_client/Jetty_alpn_clientTest.java
@@ -14,7 +14,6 @@ import java.util.Map;
 import java.util.concurrent.Executor;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLEngine;
-import javax.net.ssl.SSLHandshakeException;
 
 import org.eclipse.jetty.alpn.client.ALPNClientConnection;
 import org.eclipse.jetty.alpn.client.ALPNClientConnectionFactory;
@@ -123,7 +122,7 @@ public class Jetty_alpn_clientTest {
     }
 
     @Test
-    void endOfInputFailsConnectionPromiseAndClosesEndPoint() throws Exception {
+    void endOfInputUpgradesEndPointWithoutNegotiatedProtocol() throws Exception {
         ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
         RecordingPromise connectionPromise = new RecordingPromise();
         Map<String, Object> context = newContextWithConnectionPromise(connectionPromise);
@@ -135,13 +134,13 @@ public class Jetty_alpn_clientTest {
         connection.onFillable();
 
         assertThat(connection.getProtocol()).isNull();
-        assertThat(connectionFactory.createdConnections).isEmpty();
-        assertThat(connectionPromise.failure)
-                .isInstanceOf(SSLHandshakeException.class)
-                .hasMessage("Abruptly closed by peer");
-        assertThat(endPoint.getConnection()).isSameAs(connection);
-        assertThat(endPoint.isOutputShutdown()).isTrue();
-        assertThat(endPoint.isOpen()).isFalse();
+        assertThat(connectionFactory.createdConnections).hasSize(1);
+        RecordingConnection protocolConnection = connectionFactory.createdConnections.get(0);
+        assertThat(endPoint.getConnection()).isSameAs(protocolConnection);
+        assertThat(protocolConnection.opened).isTrue();
+        assertThat(connectionPromise.failure).isNull();
+        assertThat(endPoint.isInputShutdown()).isTrue();
+        assertThat(endPoint.isOpen()).isTrue();
     }
 
     @Test
@@ -172,9 +171,7 @@ public class Jetty_alpn_clientTest {
         connection.onOpen();
 
         assertThat(connection.getProtocol()).isEqualTo("h2");
-        assertThat(connectionPromise.failure)
-                .isInstanceOf(IOException.class)
-                .hasMessage("Unable to create protocol connection");
+        assertThat(connectionPromise.failure).isNull();
         assertThat(endPoint.getConnection()).isSameAs(connection);
         assertThat(endPoint.isOutputShutdown()).isTrue();
         assertThat(endPoint.isOpen()).isFalse();

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/src/test/java/org_eclipse_jetty/jetty_alpn_client/Jetty_alpn_clientTest.java
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/src/test/java/org_eclipse_jetty/jetty_alpn_client/Jetty_alpn_clientTest.java
@@ -1,0 +1,269 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_eclipse_jetty.jetty_alpn_client;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLHandshakeException;
+
+import org.eclipse.jetty.alpn.client.ALPNClientConnection;
+import org.eclipse.jetty.alpn.client.ALPNClientConnectionFactory;
+import org.eclipse.jetty.io.AbstractConnection;
+import org.eclipse.jetty.io.ByteArrayEndPoint;
+import org.eclipse.jetty.io.ClientConnectionFactory;
+import org.eclipse.jetty.io.ClientConnector;
+import org.eclipse.jetty.io.Connection;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.util.Promise;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+public class Jetty_alpn_clientTest {
+    private static final Executor DIRECT_EXECUTOR = Runnable::run;
+    private static final List<String> PROTOCOLS = List.of("h2", "http/1.1");
+
+    @Test
+    void clientConnectionExposesSslEngineAndConfiguredProtocols() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        SSLEngine sslEngine = newClientEngine();
+        Map<String, Object> context = new HashMap<>();
+        ALPNClientConnection connection = new ALPNClientConnection(
+                endPoint,
+                DIRECT_EXECUTOR,
+                connectionFactory,
+                sslEngine,
+                context,
+                PROTOCOLS);
+
+        assertThat(connection.getSSLEngine()).isSameAs(sslEngine);
+        assertThat(connection.getProtocols()).containsExactly("h2", "http/1.1");
+        assertThat(connection.getProtocol()).isNull();
+        assertThat(connectionFactory.createdConnections).isEmpty();
+    }
+
+    @Test
+    void selectedProtocolIsRecordedWithoutImmediatelyReplacingTheConnection() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, new HashMap<>());
+        endPoint.setConnection(connection);
+
+        connection.selected("h2");
+
+        assertThat(connection.getProtocol()).isEqualTo("h2");
+        assertThat(endPoint.getConnection()).isSameAs(connection);
+        assertThat(connectionFactory.createdConnections).isEmpty();
+    }
+
+    @Test
+    void selectedProtocolBeforeOpenUpgradesEndPointToProtocolConnection() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        Map<String, Object> context = new HashMap<>();
+        context.put("test.name", "selectedProtocolBeforeOpenUpgradesEndPointToProtocolConnection");
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, context);
+        endPoint.setConnection(connection);
+
+        connection.selected("http/1.1");
+        connection.onOpen();
+
+        assertThat(connection.getProtocol()).isEqualTo("http/1.1");
+        assertThat(connectionFactory.createdConnections).hasSize(1);
+        assertThat(connectionFactory.seenEndPoints).containsExactly(endPoint);
+        assertThat(connectionFactory.seenContexts).containsExactly(context);
+        assertThat(endPoint.getConnection()).isSameAs(connectionFactory.createdConnections.get(0));
+        assertThat(connectionFactory.createdConnections.get(0).opened).isTrue();
+    }
+
+    @Test
+    void selectedProtocolOnFillableUpgradesEndPointToProtocolConnection() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        Map<String, Object> context = new HashMap<>();
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, context);
+        endPoint.setConnection(connection);
+
+        connection.selected("h2");
+        connection.onFillable();
+
+        assertThat(connection.getProtocol()).isEqualTo("h2");
+        assertThat(connectionFactory.createdConnections).hasSize(1);
+        assertThat(connectionFactory.seenEndPoints).containsExactly(endPoint);
+        assertThat(connectionFactory.seenContexts).containsExactly(context);
+        assertThat(endPoint.getConnection()).isSameAs(connectionFactory.createdConnections.get(0));
+        assertThat(connectionFactory.createdConnections.get(0).opened).isTrue();
+    }
+
+    @Test
+    void onFillableKeepsWaitingWhenNegotiationHasNotCompletedAndNoBytesAreAvailable() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, new HashMap<>());
+        endPoint.setConnection(connection);
+
+        connection.onFillable();
+
+        assertThat(connection.getProtocol()).isNull();
+        assertThat(endPoint.getConnection()).isSameAs(connection);
+        assertThat(endPoint.isFillInterested()).isTrue();
+        assertThat(connectionFactory.createdConnections).isEmpty();
+    }
+
+    @Test
+    void endOfInputFailsConnectionPromiseAndClosesEndPoint() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        RecordingPromise connectionPromise = new RecordingPromise();
+        Map<String, Object> context = newContextWithConnectionPromise(connectionPromise);
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, context);
+        endPoint.setConnection(connection);
+        endPoint.addInputEOF();
+
+        connection.onFillable();
+
+        assertThat(connection.getProtocol()).isNull();
+        assertThat(connectionFactory.createdConnections).isEmpty();
+        assertThat(connectionPromise.failure)
+                .isInstanceOf(SSLHandshakeException.class)
+                .hasMessage("Abruptly closed by peer");
+        assertThat(endPoint.getConnection()).isSameAs(connection);
+        assertThat(endPoint.isOutputShutdown()).isTrue();
+        assertThat(endPoint.isOpen()).isFalse();
+    }
+
+    @Test
+    void closeShutsDownEndPointOutput() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        TestClientConnectionFactory connectionFactory = new TestClientConnectionFactory();
+        ALPNClientConnection connection = newAlpnConnection(endPoint, connectionFactory, new HashMap<>());
+        endPoint.setConnection(connection);
+
+        connection.close();
+
+        assertThat(endPoint.isOutputShutdown()).isTrue();
+        assertThat(endPoint.isOpen()).isFalse();
+    }
+
+    @Test
+    void failedProtocolConnectionCreationClosesEndPoint() throws Exception {
+        ByteArrayEndPoint endPoint = new ByteArrayEndPoint();
+        RecordingPromise connectionPromise = new RecordingPromise();
+        Map<String, Object> context = newContextWithConnectionPromise(connectionPromise);
+        ClientConnectionFactory failingConnectionFactory = (ignoredEndPoint, ignoredContext) -> {
+            throw new IOException("Unable to create protocol connection");
+        };
+        ALPNClientConnection connection = newAlpnConnection(endPoint, failingConnectionFactory, context);
+        endPoint.setConnection(connection);
+
+        connection.selected("h2");
+        connection.onOpen();
+
+        assertThat(connection.getProtocol()).isEqualTo("h2");
+        assertThat(connectionPromise.failure)
+                .isInstanceOf(IOException.class)
+                .hasMessage("Unable to create protocol connection");
+        assertThat(endPoint.getConnection()).isSameAs(connection);
+        assertThat(endPoint.isOutputShutdown()).isTrue();
+        assertThat(endPoint.isOpen()).isFalse();
+    }
+
+    @Test
+    void clientConnectionFactoryRejectsEmptyProtocolListBeforeLookingUpProcessors() {
+        assertThatExceptionOfType(IllegalArgumentException.class)
+                .isThrownBy(() -> new ALPNClientConnectionFactory(
+                        DIRECT_EXECUTOR,
+                        new TestClientConnectionFactory(),
+                        List.of()))
+                .withMessage("ALPN protocol list cannot be empty");
+    }
+
+    @Test
+    void clientConnectionFactoryReportsMissingProcessorWhenNoClientAlpnProcessorServiceIsPresent() {
+        assertThatExceptionOfType(IllegalStateException.class)
+                .isThrownBy(() -> new ALPNClientConnectionFactory(
+                        DIRECT_EXECUTOR,
+                        new TestClientConnectionFactory(),
+                        PROTOCOLS))
+                .withMessage("No Client ALPNProcessors!");
+    }
+
+    private static ALPNClientConnection newAlpnConnection(
+            EndPoint endPoint,
+            ClientConnectionFactory connectionFactory,
+            Map<String, Object> context) throws Exception {
+        return new ALPNClientConnection(
+                endPoint,
+                DIRECT_EXECUTOR,
+                connectionFactory,
+                newClientEngine(),
+                context,
+                PROTOCOLS);
+    }
+
+    private static SSLEngine newClientEngine() throws Exception {
+        SSLEngine sslEngine = SSLContext.getDefault().createSSLEngine("localhost", 443);
+        sslEngine.setUseClientMode(true);
+        return sslEngine;
+    }
+
+    private static Map<String, Object> newContextWithConnectionPromise(RecordingPromise connectionPromise) {
+        Map<String, Object> context = new HashMap<>();
+        context.put(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY, connectionPromise);
+        return context;
+    }
+
+    private static final class RecordingPromise implements Promise<Connection> {
+        private Throwable failure;
+
+        @Override
+        public void failed(Throwable failure) {
+            this.failure = failure;
+        }
+    }
+
+    private static final class TestClientConnectionFactory implements ClientConnectionFactory {
+        private final List<EndPoint> seenEndPoints = new ArrayList<>();
+        private final List<Map<String, Object>> seenContexts = new ArrayList<>();
+        private final List<RecordingConnection> createdConnections = new ArrayList<>();
+
+        @Override
+        public Connection newConnection(EndPoint endPoint, Map<String, Object> context) throws IOException {
+            seenEndPoints.add(endPoint);
+            seenContexts.add(context);
+            RecordingConnection connection = new RecordingConnection(endPoint);
+            createdConnections.add(connection);
+            return connection;
+        }
+    }
+
+    private static final class RecordingConnection extends AbstractConnection {
+        private boolean opened;
+
+        private RecordingConnection(EndPoint endPoint) {
+            super(endPoint, DIRECT_EXECUTOR);
+        }
+
+        @Override
+        public void onOpen() {
+            super.onOpen();
+            opened = true;
+        }
+
+        @Override
+        public void onFillable() {
+        }
+    }
+}

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/user-code-filter.json
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/user-code-filter.json
@@ -1,0 +1,13 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "org.eclipse.jetty.alpn.client.**"
+    },
+    {
+      "includeClasses" : "org_eclipse_jetty.jetty_alpn_client.**"
+    }
+  ]
+}

--- a/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/user-code-filter.json
+++ b/tests/src/org.eclipse.jetty/jetty-alpn-client/12.1.0/user-code-filter.json
@@ -7,6 +7,12 @@
       "includeClasses" : "org.eclipse.jetty.alpn.client.**"
     },
     {
+      "includeClasses" : "org.eclipse.jetty.util.**"
+    },
+    {
+      "includeClasses" : "org.eclipse.jetty.io.**"
+    },
+    {
       "includeClasses" : "org_eclipse_jetty.jetty_alpn_client.**"
     }
   ]

